keras_nlp/models/deberta/deberta_encoder.py

```diff
@@ -0,0 +1,236 @@
+# Copyright 2022 The KerasNLP Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""DeBERTa encoder block implementation based on `keras.layers.Layer`."""
+
+from tensorflow import keras
+
+from keras_nlp.models.deberta.disentangled_self_attention import (
+    DisentangledSelfAttention,
+)
+from keras_nlp.utils.keras_utils import clone_initializer
+
+from keras_nlp.layers.transformer_layer_utils import (  # isort:skip
+    merge_padding_and_attention_mask,
+)
+
+
+class DebertaEncoder(keras.layers.Layer):
+    """DeBERTa encoder.
+
+    This class follows the architecture of the DeBERTa encoder layer in the
+    paper
+    ["DeBERTaV3: Improving DeBERTa using ELECTRA-Style Pre-Training with Gradient-Disentangled Embedding Sharing"](https://arxiv.org/abs/2111.09543).
+    Users can instantiate multiple instances of this class to stack up a
+    DeBERTa encoder model.
+
+    `DebertaEncoder` is similar to `keras.layers.TransformerEncoder`, except for
```

Comment:
If we modularized the attention in `TransformerEncoder`, could we pass `DisentangledSelfAttention` in place of `MultiHeadAttention` and remove this class?

Reply:
Hmmm, this is something that was considered. We wanted to keep this layer separate (and not modify `TransformerEncoder`) because there are a few other changes we might have to make, such as passing `rel_embeddings`. Secondly, this layer is fairly limited in terms of input shapes - for example, MHA allows multiple batch dimensions `(b1, b2, other dims)` whereas disentangled allows only one `(b1, other dims)`.